### PR TITLE
Add test that newly introduced build-time fields for a min_stack for …

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
related to #2251 

## Summary
Build-time fields create changes to rule hashes without being accounted for explicitly within the version lock (essentially creating a "soft" lock). This is problematic because it creates a buggy situation where non-latest rules version bump every time due to the delta hash resulting from the differences from the build time fields. To address this, build time field introduction requires that all fields.

This adds a test to enforce that on any defined build time rule

### CODE REVIEW

File: tests/test_all_rules.py [hunk 1]
```diff
@@ -696,3 +696,29 @@ def test_rule_backports_for_restricted_fields(self):
...
+            errors = []
+            for build_field, field_versions in build_fields.items():
+                start_ver, end_ver = field_versions
+                if start_ver is not None and not Version(min_stack) >= start_ver:
+                    errors.append(f'{build_field} >= {start_ver}')
+
+            if errors:
+                err_str = ', '.join(errors)
+                invalids.append(f'{self.rule_str(rule)} uses a rule type with build fields requiring min_stack_versions'
+                                f'to be set: {err_str}')
```
Comment (Mikaayenson) [inline tests/test_all_rules.py:721]:
```suggestion
                                f' to be set: {err_str}')
```

File: tests/test_all_rules.py [hunk 2]
```diff
@@ -696,3 +696,29 @@ def test_rule_backports_for_restricted_fields(self):
...
+
+        for rule in self.production_rules:
+            min_stack = rule.contents.metadata.min_stack_version
+            build_fields = rule.contents.data.get_build_fields()
+
+            errors = []
+            for build_field, field_versions in build_fields.items():
+                start_ver, end_ver = field_versions
+                if start_ver is not None and not Version(min_stack) >= start_ver:
+                    errors.append(f'{build_field} >= {start_ver}')
```
Comment (Mikaayenson) [inline tests/test_all_rules.py:716]:
```suggestion
            if min_stack is None:
                errors.append(f'Missing min_stack_version field.')
            else:
                for build_field, field_versions in build_fields.items():
                    start_ver, end_ver = field_versions
                    if start_ver is not None and not Version(min_stack) >= start_ver:
                        errors.append(f'{build_field} >= {start_ver}')
```

**Review by Mikaayenson — APPROVED** [review]
Small suggestion
- Account for new rules that don't add min_stack_version in tool

**Review by terrancedejesus — APPROVED** [review]

File: tests/test_all_rules.py [hunk 3]
```diff
@@ -696,3 +696,29 @@ def test_rule_backports_for_restricted_fields(self):
...
+
+class TestBuildTimeFields(BaseRuleTest):
+    """Test validity of build-time fields."""
+
+    def test_build_fields_min_stack(self):
+        """Test that newly introduced build-time fields for a min_stack for applicable rules."""
+        invalids = []
+
+        for rule in self.production_rules:
+            min_stack = rule.contents.metadata.min_stack_version
```
Comment (terrancedejesus) [inline tests/test_all_rules.py:709]: Will we never use soft-forking then, only hard-forking from the metadata field??

Comment (brokensound77) [reply tests/test_all_rules.py:709]: based on our last discussions, I don't think it is supportable (or advisable)

File: tests/test_all_rules.py (hunk 1 above)
Comment (brokensound77) [reply tests/test_all_rules.py]: 01a3b71573e76f08eb98c861c8da5baf17f31add

File: tests/test_all_rules.py (hunk 2 above)
Comment (brokensound77) [reply tests/test_all_rules.py]: 01a3b71573e76f08eb98c861c8da5baf17f31add

**Comment by brokensound77** [comment]
I realized this would fail when backported to <8.3 branches and in general for branches <= build enforcement, so I added a constraint to check the current version for each build field in https://github.com/elastic/detection-rules/pull/2262/commits/b2bec63f38febd2788c07086428770e44e73f2d2

**Pull request merged**